eduTools (Features for education contexts): Waiting room; disable chat;...

ann0see opened the pull request:
DISCUSSION ONLY. Please do not merge this.

This PR includes the changes of my fork (eduTools branch) with some features added having educational needs in mind. (If we add some features of this, we can of course port them)

1. Support chat commands
2. Support waiting room (just hacked a simple way of authentication together)
3. Add support for disabling chat via chat commands
4. Do not automatically log IPs (= hide them on the server will probably not be ported to upstream. See #825)

I know that
@Brian-McBride and @gene96817 might be interested in this. 

Related to #299 #434

pljones commented:
There's little point hiding IP addresses in the Jamulus log.  Routers and firewalls likely both log them.

ann0see commented:
> There's little point hiding IP addresses in the Jamulus log. Routers and firewalls likely both log them.

Yes sure. I think we had this discussion somewhere. This PR is basically just a diff and starting point for collecting ideas concerning the education topic.

gene96817 commented:
@ann0see   Where and how do you want these topics discussed?

ann0see commented:
I would like to start a discussion on how to make Jamulus more education friendly. I know that there are discussions about adding passwords etc. but my way of "solving" the issue was that I implemented a waiting room feature. Surely some of the ideas/code can be included in some way in further projects etc. That's why I opened this PR: Basically discussing my changes and ways of implementing the waiting room, if you could/should include some here, what could be improved,...

pljones commented:
The first questions to be asked for any new feature: "Why inside Jamulus?  Can it not be done as well if not better elsewhere?  How does putting it inside Jamulus make it significantly more useful than the alternatives?"

ann0see commented:
Yes. I plan to add a few review comments here and maybe clean the history.

gene96817 commented:
I'd like to avoid handling a series of single-purpose features. It would be helpful to identify clusters of similar or related requirements.

I believe the desire for waiting room, private sessions, scheduling private sessions, and authentication can/should be consolidated into a common solution, I believe these should be outside of core Jamulus. I think of these as connection/admission management. I also would like to see the needs behind these features implemented to encourage servers to be public so that the unscheduled or unused tine is available to the community. (For example, my server utilization is less than 5%. I was delighted that a band was using it because it was the only server within 3000km. I can have a reason to make the server private. That would be a loss to the community.)

ann0see commented on src/server.cpp:
```diff
@@ -1135,6 +1181,13 @@ void CServer::MixEncodeTransmitData ( const int iChanCnt,
+            // check if client is blocked.
+            // one can always hear himself --> if we currently iterate through the blocked client we do send him audio
+            if ( ( j != iCurChanID && vecChannels[j].IsBlocked() ) || ( j != iCurChanID && currentClientBlocked ) )
+            {
```
This is the interesting part: 

In short: 

If you're blocked, nobody can hear you and you can not be heared. You will only be able to hear yourself. This makes it possible to have real private sessions in terms of "although you're connected to a server, nobody can come into your rehearsal"

ann0see commented on src/server.cpp:
```diff
@@ -1394,20 +1464,147 @@ void CServer::CreateAndSendChatTextForAllConChannels ( const int      iCurChanID
 
+    // check if chat is disabled
+    if ( bEduModeEnabled && EduModeIsFeatureDisabled( EDU_MODE_FEATURE_CHAT ) )
+    {
```
If the chat was disabled by the admin --> send an error message and don't submit it to others.

Why? Some children might spam the chat (this also happens in Zoom etc). Currently nothing can be done against that.

ann0see commented on src/server.cpp:
```diff
@@ -1394,20 +1464,147 @@ void CServer::CreateAndSendChatTextForAllConChannels ( const int      iCurChanID
-    for ( int i = 0; i < iMaxNumChannels; i++ )
+}
+void CServer::InterpretAndExecuteChatCommand ( const int iCurChanID,
+                                               const QString& strChatText )
```
A function analyzing incoming chat messages and basically enabling the administrative functions

ann0see commented on src/server.cpp:
```diff
@@ -1394,20 +1464,147 @@ void CServer::CreateAndSendChatTextForAllConChannels ( const int      iCurChanID
+    strChatCmd.remove ( 0, 3 );
+    QString ChanName = vecChannels[iCurChanID].GetName(); // to greet the user
+
+    if ( strChatCmd == strEduModePassword && bEduModeEnabled )
```
Very insecure way of authentication (plain text,...).

ann0see commented on src/server.cpp:
```diff
@@ -1394,20 +1464,147 @@ void CServer::CreateAndSendChatTextForAllConChannels ( const int      iCurChanID
+        vecChannels[iCurChanID].CreateChatTextMes ( "Enabled Chat." );
+    }
+    else if ( strChatCmd.startsWith ( "bl" ) && vecChannels[iCurChanID].IsAdmin() )
+    {   // block user
```
Manually block a user --> do not mix him into the mix of others & do not let him hear the others.

ann0see commented on src/server.cpp:
```diff
@@ -1394,20 +1464,147 @@ void CServer::CreateAndSendChatTextForAllConChannels ( const int      iCurChanID
+        vecChannels[iCurChanID].CreateChatTextMes ( "Enabled waiting room. New users will not be able to hear others." );
+    }
+    else if ( strChatCmd.startsWith( "disableWaitingRoom" ) && vecChannels[iCurChanID].IsAdmin() )
+    {
```
It should be possible to disable the waitingroom.

ann0see reviewed (commented):
Some comments on the code.

Why should some kind of moderation functions be part of Jamulus?

1. Every video conferencing tool supports features like this. Why? They have been abused, people behave badly,... 
2. Some educational institutions might block Jamulus because these features are missing
3. There's no other way I know in Jamulus to have a real private rehearsal. Although private servers provide some kind of protection, it's security through obscurity. If somebody shares the IP, everybody could join (= Zoom bombing but in Jamulus). Nowadays, in educational context, there must be some way of ensuring that children are safe and nobody unknown or unwanted can join.

So what's the intention behind these features?

It should be possible to have fully private sessions (= ensure that nobody else can join or listen to your rehearsal. This is needed in some places where children are involved). If we can solve this on public servers: This would be the best solution.

gene96817 commented on src/main.cpp:
```diff
@@ -140,6 +143,19 @@ int main ( int argc, char** argv )
 
 
+        // Log whole IP-Adresses on server ----------------------------------------------------
+        if ( GetFlagArgument ( argv,
```
What does the IP address do for us? It is possible to create a proxy for IP that is anonymized.
Some form of IP address could be needed for diagnostics.

ann0see replied:
> Some form of IP address could be needed for diagnostics.

Agree. The problem was already discussed in #576.

This can of course be removed (and probably will). I personally would like to have this (optional) function in Jamulus since it would then allow the server admin to decide if he wants to hide IPs or not. For me, this might still be a privacy issue.

gene96817 commented:
How willing are we (the developer community) willing to use DTLS for secure UDP sessions? This could be part of a secure architecture. This would eliminate spoofing and hijacking of a session.

pljones commented:
OK, fundamental to this, as far as I can see, are two functions:
- similar to the licence, prevent a connection taking part on the server
Ideally, it would work the same way - display the chat message and block further action - just without the pop-up to accept the licence enabling access
- allow a certain keyword to be used to cause "magic" to happen - i.e. if the keyword is used the rest of the command can enable or disable a user's access

So on connecting, the "owner" would need to "magic enable owner" to enable themselves and then "magic enable other1" to enable someone who has a name of "other1".  Of course, if they are one of three "NoName" users, either enable isn't supported or their IP and port number would be needed.

Similarly, "owner" will be able to incant "magic disable other2" to prevent further use of the server.

Finally, you're suggesting two more "broad" commands "magic open" to prevent the need for "magic enable whoever" and "magic close" to re-enable it.

The "magic" keyword would be configurable at server start up.

pljones commented:
So, the trick.  We've said anyone connecting is blocked unless someone issues the magic command.  However, the first person to connect _is blocked_ and that means we need to be processing _everyone's_ chat incoming, in case it's the magic command to enable yourself.  Which means the client needs to be sending them.  I'm not sure it is when the licence is what's blocking.  But it could - just they'd not get echoed with this change, as the server would need to know to ignore the client.

ann0see commented:
> similar to the licence, prevent a connection taking part on the server
Ideally, it would work the same way - display the chat message and block further action - just without the pop-up to accept the licence enabling access


Yes. That's the idea behind this PR and way of solving the problem. 

After having thought a bit more about this, another option would simply be a button or setting which disables mixing your audio in the mix of clients you select (and maybe even hide them from the mixer board.) basically a forced mute on your side. Ideally, you could say "do not mix my audio unless I select that this person can hear me" and this state could be saved client side, like the levels. This would be a small, incremental feature which at least adds some protection.

gene96817 commented:
I'd like to suggest a different framework.
Consider adding a "participation manager" based on OpenPAM. The logic and features of the participation manager is outside of Jamulus, although we could provide a very simple sample. There messages between Jamulus and the participation manager/OpenPAM would be:
1- user request (could be <null>, a string, or a n-tuple) to OpenPAM
2- message for display to user from OpenPAM
3- additional user response to OpenPAM
4- connection accept/reject from OpenPAM to Jamulus

The participation manager, at the simplest, is just a list of credentials (or UID). 
A more complicated participation manager would receive an n-tuple  (e.g. jam session name, UID). It could also make decisions based on time-of-day. All of this is managed by a different UI separate from Jamulus. Jamulus server owners can create simple or complicated systems independent of the Jamulus code.

This scheme can be configured to meet the different scenarios requested for private sessions and education applications.

ann0see commented:
I've just created a branch on this repo containing the changes of this PR so everybody can edit it. I will therefore close this PR now.

ann0see commented:
Since I deleted the branch a while ago (not sure why), I've rebased and pushed it again here: https://github.com/jamulussoftware/jamulus/tree/feature/a0/eduTools

I think with JSON-RPC at least some of the thoughts from this branch can be reused.
